receiver/memcachedreceiver/scraper.go

```diff
@@ -69,38 +65,130 @@ func (r *memcachedScraper) scrape(_ context.Context) (pdata.Metrics, error) {
+				attributes.Insert(metadata.A.Operation, pdata.NewAttributeValueString("increment"))
+				statSlice := stats.Stats
+				hits := parseFloat(statSlice["incr_hits"])
+				misses := parseFloat(statSlice["incr_misses"])
+				if hits+misses > 0 {
+					addToMetric(hitRatio, attributes, (hits / (hits + misses) * 100), now)
+				} else {
+					addToMetric(hitRatio, attributes, 0, now)
+				}
+				attributes.Insert(metadata.A.Type, pdata.NewAttributeValueString("hit"))
+				addToMetric(operationCount, attributes, parseInt(v), now)
+			case "incr_misses":
+				attributes.Insert(metadata.A.Operation, pdata.NewAttributeValueString("increment"))
+				attributes.Insert(metadata.A.Type, pdata.NewAttributeValueString("miss"))
+				addToMetric(operationCount, attributes, parseInt(v), now)
+			case "decr_hits":
+				attributes.Insert(metadata.A.Operation, pdata.NewAttributeValueString("decrement"))
+				statSlice := stats.Stats
+				hits := parseFloat(statSlice["decr_hits"])
+				misses := parseFloat(statSlice["decr_misses"])
+				if hits+misses > 0 {
+					addToMetric(hitRatio, attributes, (hits / (hits + misses) * 100), now)
+				} else {
+					addToMetric(hitRatio, attributes, 0, now)
+				}
+				attributes.Insert(metadata.A.Type, pdata.NewAttributeValueString("hit"))
+				addToMetric(operationCount, attributes, parseInt(v), now)
+			case "decr_misses":
+				attributes.Insert(metadata.A.Operation, pdata.NewAttributeValueString("decrement"))
+				attributes.Insert(metadata.A.Type, pdata.NewAttributeValueString("miss"))
+				addToMetric(operationCount, attributes, parseInt(v), now)
+			case "rusage_system":
+				attributes.Insert(metadata.A.UsageType, pdata.NewAttributeValueString("system"))
+				addToMetric(rUsage, attributes, parseFloat(v), now)
+			case "rusage_user":
+				attributes.Insert(metadata.A.UsageType, pdata.NewAttributeValueString("user"))
+				addToMetric(rUsage, attributes, parseFloat(v), now)
 			}
 		}
 	}
 
 	return md, nil
 }
 
-func addIntGauge(metrics pdata.MetricSlice, initFunc func(pdata.Metric), now pdata.Timestamp, value int64) {
-	metric := metrics.AppendEmpty()
-	initFunc(metric)
-	dp := metric.Gauge().DataPoints().AppendEmpty()
-	dp.SetTimestamp(now)
-	dp.SetIntVal(value)
+func initMetric(ms pdata.MetricSlice, mi metadata.MetricIntf) pdata.Metric {
+	m := ms.AppendEmpty()
+	mi.Init(m)
+	return m
 }
 
-func addIntSum(metrics pdata.MetricSlice, initFunc func(pdata.Metric), now pdata.Timestamp, value int64) {
-	metric := metrics.AppendEmpty()
-	initFunc(metric)
-	dp := metric.Sum().DataPoints().AppendEmpty()
-	dp.SetTimestamp(now)
-	dp.SetIntVal(value)
+// addToMetric adds a datapoint to a NumberDataPointSlice
+func addToMetric(metric pdata.NumberDataPointSlice, attributes pdata.AttributeMap, value interface{}, ts pdata.Timestamp) {
+	dataPoint := metric.AppendEmpty()
+	dataPoint.SetTimestamp(ts)
+	switch typedVal := value.(type) {
+	case int64:
+		dataPoint.SetIntVal(typedVal)
+	case float64:
+		dataPoint.SetDoubleVal(typedVal)
+	}
```

Comment:
This is a bit risky because it's very easy for a developer to pass in a value of other types. Then what happens? (Are you sure this isn't an issue with your current implementation?)

I suggest separate methods to ensure correct types.
```
addToIntMetric(metric pdata.NumberDataPointSlice, attributes pdata.AttributeMap, int64 interface{}, ts pdata.Timestamp) {
addToFloatMetric(metric pdata.NumberDataPointSlice, attributes pdata.AttributeMap, float64 interface{}, ts pdata.Timestamp) {
```

Reply:
Updated